contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,316 @@
+        (uint256 qx, uint256 qy) = recovery(h, v, r, s);
+        return getAddress(qx, qy);
+    }
+
+    /**
+     * @dev derivate public key
+     * @param privateKey - private key
+     */
+    function getPublicKey(uint256 privateKey) internal view returns (uint256, uint256) {
+        (uint256 x, uint256 y, uint256 z) = _jMult(GX, GY, 1, privateKey);
+        return _affineFromJacobian(x, y, z);
+    }
+
+    /**
+     * @dev Hash public key into an address
+     * @param qx - public key coordinate X
+     * @param qy - public key coordinate Y
+     */
+    function getAddress(uint256 qx, uint256 qy) internal pure returns (address result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0x00, qx)
+            mstore(0x20, qy)
+            result := keccak256(0x00, 0x40)
+        }
+    }
+
+    /**
+     * @dev check if a point is on the curve.
+     */
+    function isOnCurve(uint256 x, uint256 y) internal pure returns (bool result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            let p := P
+            let lhs := mulmod(y, y, p)
+            let rhs := addmod(mulmod(addmod(mulmod(x, x, p), A, p), x, p), B, p)
+            result := eq(lhs, rhs)
+        }
+    }
+
+    /**
+     * @dev Reduce from jacobian to affine coordinates
+     * @param jx - jacobian coordinate x
+     * @param jy - jacobian coordinate y
+     * @param jz - jacobian coordinate z
+     * @return ax - affine coordinate x
+     * @return ay - affine coordinate y
+     */
+    function _affineFromJacobian(uint256 jx, uint256 jy, uint256 jz) private view returns (uint256 ax, uint256 ay) {
+        if (jz == 0) return (0, 0);
+        uint256 zinv = _invModP(jz);
+        uint256 zzinv = mulmod(zinv, zinv, P);
+        uint256 zzzinv = mulmod(zzinv, zinv, P);
+        ax = mulmod(jx, zzinv, P);
+        ay = mulmod(jy, zzzinv, P);
+    }
+
+    /**
+     * @dev Point addition on the jacobian coordinates
+     * https://en.wikibooks.org/wiki/Cryptography/Prime_Curve/Jacobian_Coordinates
+     */
+    function _jAdd(
+        uint256 x1,
+        uint256 y1,
+        uint256 z1,
+        uint256 x2,
+        uint256 y2,
+        uint256 z2
+    ) private pure returns (uint256 x3, uint256 y3, uint256 z3) {
+        if (z1 == 0) {
+            return (x2, y2, z2);
+        }
+        if (z2 == 0) {
+            return (x1, y1, z1);
+        }
+        /// @solidity memory-safe-assembly
+        assembly {
+            let p := P
+            let zz1 := mulmod(z1, z1, p) // zz1 = z1²
+            let zz2 := mulmod(z2, z2, p) // zz2 = z2²
+            let u1 := mulmod(x1, zz2, p) // u1 = x1*z2²
+            let u2 := mulmod(x2, zz1, p) // u2 = x2*z1²
+            let s1 := mulmod(y1, mulmod(zz2, z2, p), p) // s1 = y1*z2³
+            let s2 := mulmod(y2, mulmod(zz1, z1, p), p) // s2 = y2*z1³
+            let h := addmod(u2, sub(p, u1), p) // h = u2-u1
```

Comment:
Similar to the `_jDouble` function, I see a couple of operations not followed from the reference here:

```
...
 if (U1 == U2)
   if (S1 != S2)
     return POINT_AT_INFINITY
   else 
     return POINT_DOUBLE(X1, Y1, Z1)
...
```

I want to make sure I understand why this is being ignored

Reply:
I believe that this is a similar case to https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4881#discussion_r1579436787, but I'm not 100% sure, so lets explore that

Reply:
if `u1=u2`, then `h = u2-u1 = 0`, and so `h*z1*z2 = 0` ... so it appears we return a point at infinity

Reply:
> if u1=u2, then h = u2-u1 = 0, and so h*z1*z2 = 0 ... so it appears we return a point at infinity

Yes, now this is clear, thanks

> with r = s2-s1, which is 0 if s2=s1 (case where the ref says we should return double) and not 0 if s2 != s1 (case where the ref says we should return point at infinity, which we do)

Right, so basically where the reference says that we should return double, it'll be doubling a point where `Z1 == 0`, which would the point at infinity anyway.